src/index.rs

```diff
@@ -399,6 +400,67 @@ impl Index {
 
         Ok(Query { inner: query })
     }
+
+    /// Parse a query leniently.
+    ///
+    /// This variant parses invalid query on a best effort basis. If some part of the query can't
+    /// reasonably be executed (range query without field, searching on a non existing field,
+    /// searching without precising field when no default field is provided...), they may get turned
+    /// into a "match-nothing" subquery.
+    ///
+    /// Args:
+    ///     query: the query, following the tantivy query language.
+    ///     default_fields_names (List[Field]): A list of fields used to search if no
+    ///         field is specified in the query.
+    ///
+    /// Returns a tuple containing the parsed query and a list of errors.
+    ///
+    /// Raises ValueError if a field in `default_field_names` is not defined or marked as indexed.
+    #[pyo3(signature = (query, default_field_names = None))]
+    pub fn parse_query_lenient(
+        &self,
+        query: &str,
+        default_field_names: Option<Vec<String>>,
+    ) -> PyResult<(Query, Vec<PyObject>)> {
+        let mut default_fields = vec![];
+        let schema = self.index.schema();
+
+        if let Some(default_field_names_vec) = default_field_names {
+            for default_field_name in &default_field_names_vec {
+                if let Ok(field) = schema.get_field(default_field_name) {
+                    let field_entry = schema.get_field_entry(field);
+                    if !field_entry.is_indexed() {
+                        return Err(exceptions::PyValueError::new_err(
+                            format!(
+                            "Field `{default_field_name}` is not set as indexed in the schema."
+                        ),
+                        ));
+                    }
+                    default_fields.push(field);
+                } else {
+                    return Err(exceptions::PyValueError::new_err(format!(
+                        "Field `{default_field_name}` is not defined in the schema."
+                    )));
+                }
+            }
+        } else {
+            for (field, field_entry) in self.index.schema().fields() {
+                if field_entry.is_indexed() {
+                    default_fields.push(field);
+                }
+            }
+        }
```

Comment:
Out of mainly curiosity, I looked at removing the mutability of `default_fields` by returning the vec from the if-expression:

```rust
    pub fn parse_query_lenient(
        &self,
        query: &str,
        default_field_names: Option<Vec<String>>,
    ) -> PyResult<(Query, Vec<PyObject>)> {
        let schema = self.index.schema();

        let default_fields = if let Some(default_field_names_vec) =
            default_field_names
        {
            default_field_names_vec
                .iter()
                .map(|field_name| {
                    schema
                        .get_field(field_name)
                        .map_err(|_err| {
                            exceptions::PyValueError::new_err(format!(
                                "Field `{field_name}` is not defined in the schema."
                            ))
                        })
                        .and_then(|field| {
                            schema.get_field_entry(field).is_indexed().then_some(field).ok_or(
                                exceptions::PyValueError::new_err(
                                    format!(
                                        "Field `{field_name}` is not set as indexed in the schema."
                                    ),
                                ))
                        })
                }).collect::<Result<Vec<_>, _>>()?
        } else {
            self.index
                .schema()
                .fields()
                .filter_map(|(f, fe)| fe.is_indexed().then_some(f))
                .collect::<Vec<_>>()
        };

```

I don't find it super compelling to make this change though. What do you think? I'm going to approve the PR regardless.

Reply:
Why not. I have a preference for functional-styling here.